File: kt/minsk/src/test/kotlin/dev/phytolizer/minsk/analysis/syntax/AssertingEnumerator.kt
package dev.phytolizer.minsk.analysis.syntax

import io.kotest.matchers.iterator.shouldBeEmpty
import io.kotest.matchers.iterator.shouldHaveNext
import io.kotest.matchers.shouldBe
import io.kotest.matchers.types.shouldBeTypeOf
import io.kotest.matchers.types.shouldNotBeTypeOf

class AssertingEnumerator(root: SyntaxNode) {
    private val _enumerator: Iterator<SyntaxNode>

    companion object {
        private fun flatten(root: SyntaxNode): List<SyntaxNode> {
            val result = mutableListOf<SyntaxNode>()
            val stack = mutableListOf(root)

            while (stack.size > 0) {
                val n = stack.removeLast()
                result.add(n)

                for (child in n.children.asReversed()) {
                    stack.add(child)
                }
            }

            return result
        }
    }

    init {
        _enumerator = flatten(root).iterator()
    }

    fun assertNode(kind: SyntaxKind) {
        _enumerator.shouldHaveNext()
        val current = _enumerator.next()
        current.kind shouldBe kind
        current.shouldNotBeTypeOf<SyntaxToken>()
    }

    fun assertToken(kind: SyntaxKind, text: String) {
        _enumerator.shouldHaveNext()
        val current = _enumerator.next()
        current.kind shouldBe kind
        val token = current.shouldBeTypeOf<SyntaxToken>()
        token.text shouldBe text
    }

    fun assertAtEnd() {
        _enumerator.shouldBeEmpty()
    }
}

File: zig/minsk_tests/src/code_analysis/syntax/AssertingEnumerator.zig
const std = @import("std");
const t = @import("framework");
const SyntaxKind = @import("minsk").code_analysis.syntax.SyntaxKind;
const SyntaxNode = @import("minsk").code_analysis.syntax.SyntaxNode;
const SyntaxToken = @import("minsk").code_analysis.syntax.SyntaxToken;

allocator: std.mem.Allocator,
nodes: []const *const SyntaxNode,
i: usize = 0,

const Self = @This();

pub fn init(allocator: std.mem.Allocator, root: *const SyntaxNode) !Self {
    const nodes = try flatten(allocator, root);
    return .{
        .allocator = allocator,
        .nodes = nodes,
    };
}

pub fn deinit(self: Self) void {
    self.allocator.free(self.nodes);
}

pub fn assertEnd(self: Self, state: *t.TestState, out_msg: *[]const u8) !void {
    try t.assert(
        state,
        self.i == self.nodes.len,
        "AssertingEnumerator disposed before reaching end",
        .{},
        out_msg,
    );
}

fn flatten(allocator: std.mem.Allocator, node: *const SyntaxNode) ![]*const SyntaxNode {
    var stack = std.ArrayList(*const SyntaxNode).init(allocator);
    defer stack.deinit();
    try stack.append(node);
    var result = std.ArrayList(*const SyntaxNode).init(allocator);
    defer result.deinit();

    while (stack.popOrNull()) |n| {
        try result.append(n);

        const children = try n.children(allocator);
        defer allocator.free(children);
        std.mem.reverse(*const SyntaxNode, children);
        try stack.appendSlice(children);
    }

    return try result.toOwnedSlice();
}

pub fn assertToken(
    self: *Self,
    state: *t.TestState,
    kind: SyntaxKind,
    text: []const u8,
    out_msg: *[]const u8,
) t.TestExit!void {
    if (!std.mem.endsWith(u8, kind.displayName(), "Token")) {
        std.debug.panic("{s} is not a token kind", .{kind.displayName()});
    }

    try t.assert(state, self.i < self.nodes.len, "assertToken() at end of list", .{}, out_msg);
    const current = self.nodes[self.i];
    self.i += 1;
    try t.assert(
        state,
        current.kind == kind,
        "assertToken: {s} != {s}",
        .{ current.kind.displayName(), kind.displayName() },
        out_msg,
    );
    const token = SyntaxNode.downcast(current, SyntaxToken);
    try t.assert(
        state,
        std.mem.eql(u8, token.text, text),
        "assertToken: '{'}' != '{'}'",
        .{ std.zig.fmtEscapes(token.text), std.zig.fmtEscapes(text) },
        out_msg,
    );
}

pub fn assertNode(
    self: *Self,
    state: *t.TestState,
    kind: SyntaxKind,
    out_msg: *[]const u8,
) t.TestExit!void {
    if (std.mem.endsWith(u8, kind.displayName(), "Token")) {
        std.debug.panic("{s} is a token kind", .{kind.displayName()});
    }

    try t.assert(state, self.i < self.nodes.len, "assertNode() at end of list", .{}, out_msg);
    const current = self.nodes[self.i];
    self.i += 1;
    try t.assert(
        state,
        current.kind == kind,
        "assertNode: {s} != {s}",
        .{ current.kind.displayName(), kind.displayName() },
        out_msg,
    );
}
